daml/ContingentClaims/Lifecycle.daml

```diff
@@ -90,49 +86,51 @@ acquire' : (Ord t, Eq a, CanAbort m)
   -- ^ the today date
   -> (AcquisitionTime t, C t a)
   -- ^ input claim and its acquisition time
-  -> m (Either (C t a) (AcquisitionTime t, C t a))
-acquire' spot t (s, (When obs c)) = do
+  -> m (AF t a)
+acquire' spot t (s, When obs c) = do
   predicate <- compare spot obs t
   if predicate then
     case obs of
-      TimeGte sNew -> pure $ Right . (max s (Time sNew), ) $ When obs c
-      other -> pure $ Right . (Time t, ) $ When (TimeGte t) c  -- if the predicate is non-deterministic, we assume that we are lifecycling at a time `t` corresponding to the first instant the predicate becomes True
+      TimeGte sNew -> pure . WhenF obs $ Right (max s (Time sNew), c)
+      other -> pure . WhenF (TimeGte t) $ Right (Time t, c)  -- if the predicate is non-deterministic, we assume that we are lifecycling at a time `t` corresponding to the first instant the predicate becomes True
     else
-      pure . Left $ When obs c
-acquire' spot today (Time s,(Cond obs c c')) = do
+      pure . WhenF obs $ Left c
+acquire' spot today (Time s, Cond obs c c') = do
   predicate <- compare spot obs s
-  if predicate 
-    then acquire' spot today (Time s,c) 
-    else acquire' spot today (Time s,c')
-acquire' spot today (MinusInfinity,(Cond obs c c')) = do
+  acquire' spot today . (Time s, ) $ if predicate then c else c'
+acquire' _ _ (MinusInfinity, Cond {}) = do
   abort "Found Cond node with undefined acquisition time"
-acquire' spot today (_,c@(Anytime obs _)) = do
+acquire' spot today (_, Anytime obs c) = do
   predicate <- compare spot obs today
-  pure $ if predicate then Right (Time today,c) else Left c  -- the acquisition time of an Anytime node is assumed to be the time at which we lifecycle
-acquire' _ _ (s,other) = pure $ Right (s, other)
+  pure . AnytimeF obs $ if predicate then Right (Time today,c) else Left c  -- the acquisition time of an Anytime node is assumed to be the time at which we lifecycle
+acquire' _ _ (s, other) = pure $ Right . (s, ) <$> project other
 
 -- | Evaluate any observables in `Scale` nodes, accumulating scale factors top-down.
 -- Log the scale factors with their corresponding leaf values. 
--- Skip `Or` and `Anytime` branches, guaranteeing liveness. 
+-- Stop recursion at `Or` and `Anytime` branches, guaranteeing liveness.
 -- Replace any Ones that can be reached with Zeros.
 lifecycle' : (Ord t, CanAbort m)
         => (a -> t -> m Decimal)
-        -> (Decimal, (AcquisitionTime t, C t a))
-        -- ^ the input claim, the acquisition time of its sub-trees and the accumulated scaling factor
+        -> (Decimal, (AcquisitionTime t, AF t a))
```

Comment:
I think this is getting to the point where it may be worth using a dedicated data type instead of `(Decimal, (AcquisitionTime t, AF t a)` for the carrier, to keep the code tidy. Don't add it to the export list though.

What do you think?

It would help with pattern matches like on line 127, where this is getting unwieldy.

Reply:
Thanks. 

I introduced some type synonyms to simplify the function signatures in my latest commit.

I have been playing around with defining an additional record type to make the pattern-matching syntax a bit cleaner, but I am not sure on the end result. 

Let me open a separate PR with the change to discuss.

Reply:
I just thought I would mention, in case you aren't familiar (we don't advertise it in Daml) that record syntax can also be used for pattern matching. The more complicated version is something like:
```
f MyData{field1=a,b,..}
```
If you're not familiar then my comment above may seem somewhat vague. Have a search in the Haskell language docs.

Reply:
Thanks! I did spent some time with Brian and Georg trying to recall how to pattern-match on a record in Daml :)